src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs

```diff
@@ -2041,13 +2041,13 @@ protected override void AfterLeftChildHasBeenVisited(BoundBinaryOperator binary)
 
             BinaryOperatorKind op = binary.OperatorKind.Operator();
 
-            // learn from non-null constant
+            // learn from non-null expression
             BoundExpression operandComparedToNonNull = null;
-            if (isNonNullConstant(binary.Left))
+            if (leftType.IsNotNull && rightType.MayBeNull && binary.Right.ConstantValue?.IsNull != true)
             {
                 operandComparedToNonNull = binary.Right;
             }
-            else if (isNonNullConstant(binary.Right))
+            else if (leftType.MayBeNull && binary.Left.ConstantValue?.IsNull != true && rightType.IsNotNull)
```

Comment:
>binary.Left.ConstantValue?.IsNull [](start = 43, length = 33)

Why don't we have to skip through nullable conversions like we were doing previously? What happens if you were to do `left?.IntProp != (int?)null`?

Reply:
It appears that nullable conversions have a ConstantValue when the value being converted is `null`, but not when the value being converted is non-null. Tracked in #24380

Reply:
Basically, if the decision gets made one way (nullable conversions from constants should have ConstantValues) this code can stay as-is. If it gets made the other way (no nullable conversion should have a ConstantValue), this code may have to change.